feat: use go.work to reduce core library dependencies

slsyy opened the pull request:
some shenaningas were introduced to run tests&lint over all modules Probably it won't be necessary, when
https://github.com/golang/go/issues/71294 is resolved

ccoVeille commented on .github/workflows/lint.yml:
```diff
@@ -8,17 +8,36 @@ on:
         with:
-          version: v1.56.2
+          version: ${{ env.GOLANGCI_LINT_VERSION }}
+          working-directory: ${{ matrix.modules }}
```
I'm unsure to understand why you are computing all modules then run sequentially via the matrix

For me, as long as you have a go.work, `go test ./...` and `golangci-lint run ./...` should work in a transverse way

So for me, you don't need the `go list -json | jq`

Could you break something dbtests/mysql_test.go either with code and style

Then try to run go test and golangci-lint

They should be detected.

It's worth trying

slsyy replied:
> For me, as long as you have a go.work, go test ./... and golangci-lint run ./... should work in a transverse way

Unfortunately it does not work in that way, you need to run `go test` in each `go.mod` directory, check this https://github.com/golang/go/issues/71294

Anyway I am still experimenting and I will try few different ideas to check, which one least sucks

ty for your input

ccoVeille replied:
For anyone interested

> This proposal has been added to the [active column](https://go.dev/s/proposal-status#active) of the proposals project
and will now be reviewed at the weekly proposal review meetings.

_Originally posted by @rsc in https://github.com/golang/go/issues/71294#issuecomment-2637865401_

🤞

ccoVeille replied:
> Unfortunately it does not work in that way, you need to run go test in each go.mod directory, check this golang/go#71294\n\nAnyway I am still experimenting and I will try few different ideas to check, which one least sucks

Hi @slsyy 

While I'm now convinced about the lack of feature in `go test`, I'm doubtful about the fact golangci-lint doesn't support it

A few weeks ago, I added a simple go.work file in gofr framework

- https://github.com/gofr-dev/gofr/pull/1426

And all the errors appeared locally when I was running golangci-lint manually locally

slsyy replied:
@ccoVeille here is a proof. I purposely entered some garbage to `dbtests` module, so lint should fail, but a simple `golangci-lint run` does not detect it

![image](https://github.com/user-attachments/assets/214ea21f-18f0-4550-902d-bad676e7dd7f)

ccoVeille commented:
For records, Go 1.25 will have `go test work` to test all submodules listed in the go.work file

cc @slsyy it was finally merged

https://github.com/golang/go/issues/50745#issuecomment-2730714430

slsyy commented:
@ccoVeille thanks a lot. I will check it

slsyy committed:
feat: use go.work to reduce core library dependencies

some shenaningas were introduced to run tests&lint over all modules
Probably it won't be necessary, when
golang/go#71294 is resolved

andreynering reviewed (approved):
Superb work @slsyy! 🚀

The users will love this! We now depend on two external packages only 🙂

Small notes before merging:

* Running `go mod tidy` on some of the dirs (like `dbtests/`) introduce small changes due to package updates.
* We might need to update the Dependabot config to look for all the module directories. I'm not sure it that would happen by default. We could try merging as well and doing a manual trigger of Dependabot to test if it's working, or configure later if needed.

slsyy commented:
> Running go mod tidy on some of the dirs (like dbtests/) introduce small changes due to package updates.

Fixed

> We might need to update the Dependabot config to look for all the module directories. I'm not sure it that would happen by default

I altered a dependabot configuration based on https://github.com/charmbracelet/x/blob/main/.github/dependabot.yml

andreynering commented:
> I altered a dependabot configuration based on https://github.com/charmbracelet/x/blob/main/.github/dependabot.yml

@slsyy Looks great! Feel free to merge once you're ready.
